controllers/kafkamanagement/kafkauser_controller.go

```diff
@@ -533,3 +557,173 @@ func (r *KafkaUserReconciler) getKafkaUserCredsFromSecret(
+		SignatureAlgorithm: x509.SHA256WithRSA,
+	}
+
+	csrBytes, _ := x509.CreateCertificateRequest(rand.Reader, &template, keyBytes)
+	strBuf := strings.Builder{}
+	err := pem.Encode(&strBuf, &pem.Block{Type: "NEW CERTIFICATE REQUEST", Bytes: csrBytes})
+	if err != nil {
+		return "", err
+	}
+
+	return strBuf.String(), nil
+}
+
+func (r *KafkaUserReconciler) UpdateCertSecret(ctx context.Context, secret *v1.Secret, certResp *v1beta1.CertificateStruct) error {
+	secret.Data["csr"] = []byte(certResp.Csr)
+	secret.Data["id"] = []byte(certResp.ID)
+	secret.Data["expiryDate"] = []byte(certResp.ExpiryDate)
+	secret.Data["signedCertificate"] = []byte(certResp.SignedCertificate)
+
+	err := r.Update(ctx, secret)
+	if err != nil {
+		return err
+	}
+	
+	return nil
+}
+
+func (r *KafkaUserReconciler) handleCertificateEvent(
+	newObj *v1beta1.KafkaUser,
+	oldCerts []*clusterresourcessv1beta1.SecretReference,
+) {
+	ctx := context.TODO()
+	l := log.FromContext(ctx)
+
+	for _, newCert := range newObj.Spec.CertificateSecretRefs {
+		var exist bool
+		for _, oldCert := range oldCerts {
+			if *newCert == *oldCert {
+				exist = true
+				break
+			}
+		}
+
+		if exist {
+			continue
+		}
+
+		err := r.handleCreateCertificate(ctx, newObj, l, newCert)
+		if err != nil {
+			l.Error(err, "Cannot create Kafka user Mtls certificate", "cert", newCert)
+			r.EventRecorder.Eventf(newObj, models.Warning, models.CreatingEvent,
+				"Cannot create user Mtls certificate. Reason: %v", err)
+		}
+		oldCerts = append(oldCerts, newCert)
+	}
+}
+
+func (r *KafkaUserReconciler) handleCreateCertificate(ctx context.Context, user *v1beta1.KafkaUser, l logr.Logger, certRef *clusterresourcessv1beta1.SecretReference) error {
+	username, _, err := r.getKafkaUserCredsFromSecret(user.Spec)
+	if err != nil {
+		l.Error(
+			err, "Cannot get Kafka user creds from secret",
+			"kafka user spec", user.Spec,
+		)
+		r.EventRecorder.Eventf(
+			user, models.Warning, models.FetchFailed,
+			"Fetch user credentials from secret is failed. Reason: %v",
+			err,
+		)
+
+		return err
+	}
+
+	certSecret := &v1.Secret{}
+	kafkaUserCertSecretNamespacedName := types.NamespacedName{
+		Name:      certRef.Name,
+		Namespace: certRef.Namespace,
+	}
+	err = r.Client.Get(ctx, kafkaUserCertSecretNamespacedName, certSecret)
+	if k8serrors.IsNotFound(err) {
+		l.Error(err, "Cannot get Kafka user Cert Secret. Secret is not found",
+			"secret name", certRef.Name,
+		)
+		r.EventRecorder.Eventf(
+			user, models.Warning, models.FetchFailed,
+			"Fetch user Cert Secret is failed. Reason: %v",
+			err,
+		)
+
+		return err
+	}
+
+	if _, exist := certSecret.Data[models.SignedCertificate]; !exist {
+		csr, err := r.GenerateCSR(certSecret)
+		if err != nil {
+			l.Error(err, "Cannot generate CSR for Kafka user certificate creation",
+				"user", user.Name,
+			)
+			r.EventRecorder.Eventf(
+				user, models.Warning, models.GenerateFailed,
+				"Generate CSR is failed. Reason: %v",
+				err,
+			)
+		}
```

Comment:
You should return the `err` here, shouldn't you? Please fix similar cases below.

Reply:
done